### ISSUE TITLE
handleSubmit swallowing errors

### ISSUE THREAD

**Issue by tume** [erikras/redux-form#2068]
I'm currently returning promise in my own handleSubmit which gets rejected on error but it seems that redux-form is catching this and handling it internally. 

Haven't been using redux-form for a long so might be thinking this totally backwards but I was hoping that I could just simple get the status of submit from the promise. Now it seem always to resolve even if it contains validation errors.

What is the best way of achieving this? should I just check by hand if it contains validation errors and return rejected promise? and currently I can't do all the logic inside my handleSubmit callback, so I return the promise for further processing.

**Comment by MilllerTime** [comment]
Hey @tume, if you're throwing errors in a promise chain, I can almost guarantee that it's the promises that are swallowing your errors, not redux-form. [Turns out promises can be tricky when it comes to exception handling](https://www.google.com/search?q=promises+swallowing+errors) - the key lies in the little details of promises. They are worth [studying closely](https://pouchdb.com/2015/05/18/we-have-a-problem-with-promises.html)!

However, if you return a promise from `handleSubmit` and throw a `SubmissionError`, it will cause your promise chain to end in a rejected promise. Redux-form will catch that, see the special `SubmissionError`, and handle validation errors. _It's up to you to determine if your server response has validation errors and to throw the `SubmissionError`_. This is by design. [The docs have a simple example.](http://redux-form.com/6.1.1/docs/api/SubmissionError.md/)

**Comment by tume** [comment]
Thanks for your explanation. Currently i'm not so interested in specific validation errors just ok / error in general. If I now understood correctly there is no way that redux-form would return rejected promise in handleSubmit no matter what I return in my own handler? It will just handle the submissionErrors and resolve it?

Just feels a bit strange that I have now errors in promise which I can't use .catch() to catch.

**Comment by MilllerTime** [comment]
Correct, redux-form's `handleSubmit` doesn't return anything you can use AFAIK. The method is just an event handler for form submission. I'm confused about what you're trying to accomplish with promises - you should have full control of the chain, including `catch`ing errors.

While you can return a promise to `handleSubmit` to make it aware of a pending request, that should be the last step in your promise chain and you have full control until that point.

As for taking action on errors, you can throw a `SubmissionError` with an `_error` key (representing a form level error), which will be passed to your decorated component as an `error` prop. You can use that to determine if submission failed from within your component. It's roughly documented here: http://redux-form.com/6.1.1/docs/api/ReduxForm.md/#-onsubmit-function-optional-

If you still think something is missing in regards to promise exception handling, could you post a simplified example showing the issue?

**Comment by WirthAware** [comment]
Hi @MilllerTime do you know whats the best way with redux-form v5 to handle this? The `SubmissionError` comes in with v6. 

I'm calling submit() on a Form ref the way like in the "submit from parent" example from the docs in v5.3.3. The submit returns a Promise which always resolves. 

In my case i submit many forms that way and want to Promise.all that form promises to check if one form has failed.

**Comment by MilllerTime** [comment]
@WirthAware It seems like `SubmissionError` was mostly added for readability. In v5, you can still reject your Promise with an error map object just like you would pass to a `SubmissionError` in v6. The special form level `_error` key should work as well. The main difference is that you must _explicitly_ reject the promise with that object instead of throwing an error that _implicitly_ rejects it.

Your decorated component should then behave the same way.

The `onSubmit` description seems to explain this here (can't link to the specific section because the old doc site uses hash history):  http://redux-form.com/5.3.3/#/api/reduxForm

**Comment by WirthAware** [comment]
@MilllerTime Thanks for your explanation. Thats working fine for me. In our project we have a kind of wizard of forms. Each form has an own submit. Thats working fine for each form. 

Now we want to have a submit for each form to submit them all. The wizard could have e.g. 4 steps. Each step is a form and now i want to submit all forms with one button. I have seen in v6.2 there is a new action to submit, Thats pretty cool! But in that project we are still on v5. 

In my solution i have a ref on each form. With that ref i could call submit which returns a promise. 

Copied from the docs: http://redux-form.com/5.3.3/#/examples/submit-from-parent?_k=7ied73

`
const myPromise = this.refs.myForm.submit()  // will return a promise
`

`myPromise` now always resolves also when `myForm` `onSubmit` is rejecting will `myPromise` resolve. I would expect when onSubmit handleSubmit will reject that `myPromise` will also rejecting. When i submit many forms i need to know if one promise has failed. But when all promises resolve the wizard will not informed about an error. 

I have a workaround that works for now. But i'm interested why `myPromise` will always resolve also when handleSubmit will reject.

**Comment by MilllerTime** [comment]
@WirthAware I think I understand now. I haven't run into the same use-case you're describing, but have you tried setting `returnRejectedSubmitPromise` to `true` in your form config? [According to the docs](http://redux-form.com/5.3.3/#/api/reduxForm) it defaults to false. I'm not sure why the default behavior is to swallow errors, but maybe that will do the trick.

**Comment by WirthAware** [comment]
@MilllerTime that does the trick! Thank you for your help!

**Comment by Dema** [comment]
I've just stumbled upon this error in slightly different aspect.

My app is a simple page with login form. Upon submit it calls ajax service that validates credentials.
If they are valid, some component is shown instead of login form. But the problem is that 
this component throws an exception in render method. I'm also using redux-promised middleware that 
catches errors and dispatches XXX_FAILED action. If I disable it, everything works.
In handleSubmit.js you have the following code:

```javascript
if (isPromise(result)) {
	startSubmit();
	return result.then(function (submitResult) {
	stopSubmit();
	setSubmitSucceeded();
	if (onSubmitSuccess) {
		onSubmitSuccess(submitResult, dispatch);
	}
	return submitResult;
	}, function (submitError) {
	var error = submitError instanceof SubmissionError ? submitError.errors : undefined;
	^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
	stopSubmit(error);
	setSubmitFailed.apply(undefined, _toConsumableArray(fields));
	if (onSubmitFail) {
		onSubmitFail(error, dispatch);
	}
	return error;
	});
} else {
```
At the highlighted stage you are checking if it is a valid rejected promise from ajax, and if it is, it's perfectly valid to dispatch
STOP_SUBMIT action in stopSubmit. But if it was an error in React itself (like my case of throwing error in render) then 
React is in inconsistent state and when STOP_SUBMIT is sent it tries to rerender my broken component and fails with the cryptic error

```
ReactReconciler.js:64 Uncaught (in promise) TypeError: Cannot read property 'getNativeNode' of null
    at Object.getNativeNode (webpack:///./~/react/lib/ReactReconciler.js?:64:28)
    at ReactCompositeComponentWrapper.getNativeNode (webpack:///./~/react/lib/ReactCompositeComponent.js?:369:28)
    at Object.getNativeNode (webpack:///./~/react/lib/ReactReconciler.js?:64:29)
    at Object.updateChildren (webpack:///./~/react/lib/ReactChildReconciler.js?:89:48)
    at ReactDOMComponent._reconcilerUpdateChildren (webpack:///./~/react/lib/ReactMultiChild.js?:189:32)
    at ReactDOMComponent._updateChildren (webpack:///./~/react/lib/ReactMultiChild.js?:288:31)
    at ReactDOMComponent.updateChildren (webpack:///./~/react/lib/ReactMultiChild.js?:276:12)
    at ReactDOMComponent._updateDOMChildren (webpack:///./~/react/lib/ReactDOMComponent.js?:883:12)
    at ReactDOMComponent.updateComponent (webpack:///./~/react/lib/ReactDOMComponent.js?:712:10)
    at ReactDOMComponent.receiveComponent (webpack:///./~/react/lib/ReactDOMComponent.js?:668:10)
```

As I understand, at the highlighted stage we should just rethrow the error like this:

```javascript
if(!error) {
	throw submitError;
}
```
If I add this code, everything works perfectly. Am I correct?

**Comment by oyeanuj** [comment]
@MilllerTime I'm facing the same issue, and your initial comment intuitively seems like the diagnosis. However, given that `returnRejectedSubmitPromise` has been taken out, I haven't been able to figure out a solution here. 

I was wondering if you were able to diagnose by looking at a simplified example (as per your above comment). Any thoughts are much appreciated.

```js
//how handleSubmit is being called:

renderDebouncedInput = (field) => {
  const { handleSubmit, onChangeSubmit } = this.props;

  return (
    <DebounceInput
      debounceTimeout = {500}
      onChange = {(e) => {
        field.input.onChange(e);
        handleSubmit((values, dispatch, props) => {
          const newValues = values;
          newValues[field.name] = e.target.value;
          onChangeSubmit(newValues, dispatch, props);
        })();
      }}
      value = {field.input.value}
    />
  );
};


//the submit function
updateField = (values, dispatch, props) => {
  return new Promise((resolve, reject) => {
    return updateField(values.fieldName)
      .then((response) => {
        if (response) { resolve(); }
      })
      .catch((response) => {
        const error = response.body && response.body.base;
        reject(new SubmissionError({ _error: "That's weird. " + error }));
      });
  });
};

```

The error I get is:
```shell
Uncaught (in promise) 
SubmissionError {
  errors: Object, 
  message: "Submit Validation Failed", 
  name: "SubmissionError", 
  stack: "SubmissionError: Submit Validation Failed
  ...
}
//error line points to the reject statement
```

I assume I am tripping up somewhere with error being swallowed given the way I am calling `handleSubmit`?

**Comment by MilllerTime** [comment]
Hey @oyeanuj, your `updateField` function works when used directly because it returns a promise that redux-form can chain onto. The anonymous function you're passing to `handleSubmit` in your `<DebounceInput>` component doesn't return any value. However, it seems like you're delegating to another function, `onChangeSubmit`. I imagine that probably returns a promise, so you could just return the result like so:
```JS
return onChangeSubmit(newValues, dispatch, props);
```
Since you're not passing the promise to redux-form, the `SubmissionError` isn't caught, and I suspect that is your error.


BTW, there is no need to wrap a promise-returning piece of code with a new promise constructor. It just adds overhead and complexity. You should be able to simply write:
```JS
const yourSubmitHandler = (values, dispatch, props) => {
  return updateField(values.fieldName)
    .catch((response) => {
      const error = response.body && response.body.base;
      throw new SubmissionError({ _error: "That's weird. " + error });
    });
};
```
By throwing the `SubmissionError`, you're implicitly rejecting the promise chain, and a resolved promise will just pass through and be returned as-is 🙂

**Comment by MilllerTime** [comment]
@Dema Shouldn't you be handling render errors within the component itself? I'm confused why a form submission handler should need to worry about that.

**Comment by oyeanuj** [comment]
@MilllerTime Your diagnosis was spot on - returning the result of `onChangeAction` (which was just `updateField` in the example above) made the thing work. So, `handleSubmit` now gets the promise or a value returned which it wasn't earlier.

Thank you for that and the additional code review, and tips! Much appreciated :)

**Comment by Dema** [comment]
You mean I should wrap all `render()` and all lifecycle methods in `try{} catch{}` block? I don't think it's a good idea.
The error was thrown deep in React internals after `render()` was finished. I've imported non-existing function from module and tried to call it in `componentWillMount()` hook. 
So it crashed and the component never got rendered.

So there is not much I can do in catch handler, it wasn't a legitimate error I could recover from. Allowing it to bubble up to the top is the best I can do, I think.

Here is the relevant thread from Dan Abramov https://twitter.com/dan_abramov/status/7988384333680640

He wrote it after I submitted a bug report about React throwing cryptic `Error: getHostNode() is null`.

The code in handleSubmit.js a bit different now, but it still suffers from the same error:
```javascript
}, function (submitError) {
var error = submitError instanceof _SubmissionError2.default ? submitError.errors : undefined;

stopSubmit(error);
^^^^^(1)
setSubmitFailed.apply(undefined, _toConsumableArray(fields));
if (onSubmitFail) {
	onSubmitFail(error, dispatch);
}
if (error || onSubmitFail) {
	// if you've provided an onSubmitFail callback, don't re-throw the error
	return error;
} else {
	throw submitError;
	^^^^^ (2)
}
});
```

It's too late to throw the error in (2), `stopSubmit` action is already dispatched in (1) and will fail because React state is garbled and `dispatch` indirectly calls `setState` that forces component to rerender but it's not rendered for the first time because of an error.
We should re-throw non-SubmitError exception before dispatching `stopSubmit` action.
https://twitter.com/dan_abramov/status/798856890813845504

**Comment by jmar777** [comment]
+1000 to not swallowing unrelated errors.

During development, my team has seen cases where after form submission, we would perform a `history.pushState()` and encounter and error in an entirely unrelated component. Because we were still within the same call stack as `handleSubmit()`, the error would be caught and swallowed.

Needless to say, that's not redux-form's fault, but it does make debugging these situations a bit nightmarish. I'm a bit agnostic as to which solution is used here, but will lend a vote towards anything that doesn't result in unrelated errors being swallowed. :)

Happy to get my hands dirty as well... I set off to make a PR when I found this issue, and realized it's not quite as straight forward as I imagined. If there's a consensus on the right approach to take, I could take a crack at it. Thanks for all the awesome work!
